src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,156 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = tail = new Node<>(null, value, null);
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        validateIndexForAdd(index);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> nodeAtIndex = getNode(index);
+        Node<T> newNode = new Node<>(nodeAtIndex.prev, value, nodeAtIndex);
+        if (index == 0) {
+            head = newNode;
+        } else {
+            nodeAtIndex.prev.next = newNode;
+        }
+        nodeAtIndex.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToRemove = getNode(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> nodeToRemove = head;
+        while (nodeToRemove != null) {
+            if (objectsEqual(object, nodeToRemove.value)) {
+                unlink(nodeToRemove);
+                return true;
+            }
+            nodeToRemove = nodeToRemove.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeFromTail(int index) {
+        Node<T> node = tail;
+        for (int i = size - 1; i > index; i--) {
+            node = node.prev;
+        }
+        return node;
+    }
+
+    private Node<T> getNode(int index) {
+        checkIndex(index);
+        return index <= size / 2 ? getNodeFromHead(index) : getNodeFromTail(index);
+    }
```

Comment:
locate `getNode` before `getNodeFromTail`

Reply:
not fixed